src/NumericalAlgorithms/LinearSolver/Gmres.hpp

```diff
@@ -0,0 +1,209 @@
+ * operator action \f$A(x)\f$ must be provided for an argument \f$x\f$.
+ *
+ * The GMRES algorithm does not require the operator \f$A\f$ to be symmetric or
+ * positive-definite. Note that other algorithms such as conjugate gradients may
+ * be more efficient for symmetric positive-definite operators.
+ */
+template <typename VarsType>
+struct Gmres {
+ private:
+  struct ConvergenceCriteria {
+    using type = Convergence::Criteria;
+    static constexpr OptionString help =
+        "Determine convergence of the algorithm";
+  };
+  struct Restart {
+    using type = size_t;
+    static constexpr OptionString help = "Iterations to run before restarting";
+    static size_t default_value() noexcept { return 0; }
+  };
+  struct Verbosity {
+    using type = ::Verbosity;
+    static constexpr OptionString help = "Logging verbosity";
+  };
+
+ public:
+  static constexpr OptionString help =
+      "A serial GMRES iterative solver for nonsymmetric linear systems of "
+      "equations.";
+  using options = tmpl::list<ConvergenceCriteria, Verbosity, Restart>;
+
+  Gmres(Convergence::Criteria convergence_criteria, ::Verbosity verbosity,
+        size_t restart = 0) noexcept
+      // clang-tidy: trivially copyable
+      : convergence_criteria_(std::move(convergence_criteria)),  // NOLINT
+        verbosity_(std::move(verbosity)),                        // NOLINT
+        restart_(restart > 0 ? restart : convergence_criteria_.max_iterations) {
+    initialize();
+  }
+
+  Gmres() = default;
+  Gmres(const Gmres& /*rhs*/) = default;
+  Gmres& operator=(const Gmres& /*rhs*/) = default;
+  Gmres(Gmres&& /*rhs*/) noexcept = default;
+  Gmres& operator=(Gmres&& /*rhs*/) noexcept = default;
+  ~Gmres() = default;
+
+  void initialize() noexcept {
+    orthogonalization_history_.reserve(restart_ + 1);
+    basis_history_.resize(restart_);
+    preconditioned_basis_history_.resize(restart_);
+  }
+
+  const Convergence::Criteria& convergence_criteria() const noexcept {
+    return convergence_criteria_;
+  }
+
+  Convergence::Reason convergence_reason() const noexcept {
+    ASSERT(convergence_reason_,
+           "Tried to retrieve the convergence reason, but has not performed a "
+           "solve yet.");
+    return *convergence_reason_;
+  }
+
+  size_t num_iterations() const noexcept { return iteration_; }
+
+  void pup(PUP::er& p) noexcept {  // NOLINT
+    p | convergence_criteria_;
+    p | verbosity_;
+    p | restart_;
+    if (p.isUnpacking()) {
+      initialize();
+    }
+  }
+
+  template <typename LinearOperator, typename SourceType,
+            typename Preconditioner = IdentityPreconditioner<VarsType>>
+  VarsType operator()(LinearOperator&& linear_operator,
+                      const SourceType& source, const VarsType& initial_guess,
+                      const Preconditioner& preconditioner =
+                          IdentityPreconditioner<VarsType>{}) const noexcept {
```

Comment:
This changes the visible state of the class, so it should not be const.

Reply:
I removed the state changing and instead return a `Convergence::HasConverged` that reports the information. Not sure why I didn't do that in the first place